wavefront_python_sdk/common/utils.py

```diff
@@ -0,0 +1,129 @@
+
+
+class AtomicCounter:
+    """An atomic, thread-safe incrementing counter.
+    # >>> counter = AtomicCounter()
+    # >>> counter.increment()
+    # 1
+    # >>> counter.increment(4)
+    # 5
+    # >>> counter = AtomicCounter(42.5)
+    # >>> counter.value
+    # 42.5
+    """
+
+    def __init__(self, initial=0):
+        self.value = initial
+        self._lock = threading.Lock()
+
+    def increment(self, num=1):
+        with self._lock:
+            self.value += num
+            return self.value
+
+
+def sanitize(s):
+    whitespace_sanitized = re.sub(r"[\s]+", "-", s)
+    if "\"" in whitespace_sanitized:
+        return "\"" + re.sub(r"[\"]+", '\\\\\"', whitespace_sanitized) + "\""
+    else:
+        return "\"" + whitespace_sanitized + "\""
+
+
+def is_blank(s):
+    return s is None or len(s) == 0 or s.isspace()
+    # return len(re.sub(r"[\s]+", "", s)) == 0
+
+
+def metric_to_line_data(name, value, timestamp, source, tags, default_source):
+    # /*
+    #  * Wavefront Metrics Data format
+    #  * <metricName> <metricValue> [<timestamp>] source=<source> [pointTags]
+    #  * Example: "new-york.power.usage 42422 1533531013 source=localhost datacenter=dc1"
+    #  */
+    if is_blank(name):
+        raise ValueError("Metrics name cannot be blank")
+
+    if is_blank(source):
+        source = default_source
+
+    str_builder = [sanitize(name), float(value)]
+    if timestamp is not None:
+        str_builder.append(int(timestamp))
+    str_builder.append("source=" + sanitize(source))
+    if tags is not None:
+        for key in tags:
+            if is_blank(key):
+                raise ValueError("Metric point tag key cannot be blank")
+            if is_blank(tags[key]):
+                raise ValueError("Metric point tag value cannot be blank")
+            str_builder.append(sanitize(key) + '=' + sanitize(tags[key]))
+    return ' '.join('{0}'.format(s) for s in str_builder) + "\n"
+
+
+def histogram_to_line_data(name, centroids, histogram_granularities, timestamp, source, tags, default_source):
+    if is_blank(name):
+        raise ValueError("Histogram name cannot be blank")
+
+    if histogram_granularities is None or len(histogram_granularities) == 0:
+        raise ValueError("Histogram granularities cannot be null or empty")
+
+    if centroids is None or len(centroids) == 0:
+        raise ValueError("A distribution should have at least one centroid")
+
+    if is_blank(source):
+        source = default_source
+
+    line_builder = []
+    for histogram_granularity in histogram_granularities:
+        str_builder = [histogram_granularity.identifier]
+        if timestamp is not None:
+            str_builder.append(int(timestamp))
+        for centroid_1, centroid_2 in centroids:
+            str_builder.append("#" + repr(centroid_2))
+            str_builder.append(centroid_1)
+        str_builder.append(sanitize(name))
+        str_builder.append("source=" + sanitize(source))
+        if tags is not None:
```

Comment:
This logic seems repeated in a few places: `metric_to_line_data`, `histogram_to_line_data` etc. Maybe move into a method and reuse.